File: scripts/algorithms/F/Find the Minimum and Maximum Number of Nodes Between Critical Points/Find the Minimum and Maximum Number of Nodes Between Critical Points.py
# Runtime: 687 ms (Top 71.4%) | Memory: 56.78 MB (Top 82.0%)

class Solution:
    def nodesBetweenCriticalPoints(self, head: Optional[ListNode]) -> List[int]:
        idx, i = [], 1
        prev, cur = head, head.next
        while cur and cur.next:
            if prev.val < cur.val > cur.next.val or prev.val > cur.val < cur.next.val:
                idx.append(i)
            prev = cur
            cur = cur.next
            i += 1

        if len(idx) < 2:
            return [-1, -1]
        
        minDist = min(j - i for i, j in pairwise(idx))
        maxDist = idx[-1] - idx[0]

        return [minDist, maxDist]

File: scripts/algorithms/F/Find the Minimum and Maximum Number of Nodes Between Critical Points/Find the Minimum and Maximum Number of Nodes Between Critical Points.cpp
// Runtime: 166 ms (Top 87.34%) | Memory: 113.60 MB (Top 78.69%)

class Solution {
   public:
    vector<int> nodesBetweenCriticalPoints(ListNode *head) {
        ListNode *curr = head, *prev = NULL;
        int minDistance = INT_MAX, maxDistance = -1, size = 0, preCP = 0,
            first = 0;
        while (curr) {
            if (curr->next && prev) {
                if ((curr->val < prev->val && curr->val < curr->next->val) ||
                    (curr->val > prev->val && curr->val > curr->next->val)) {
                    if (first == 0) {
                        first = size;
                    } else {
                        minDistance = min(minDistance, size - preCP);
                        maxDistance = max(maxDistance, size - first);
                    }
                    preCP = size;
                }
            }
            size++;
            prev = curr;
            curr = curr->next;
        }
        return {minDistance == INT_MAX ? -1 : minDistance, maxDistance};
    }
};
